### PR TITLE
Avoid purging Revisions in use by third-party packages

### LINKED ISSUE #10678: Avoid purging Revisions in use by third-party packages
I originally raised this as part of some rework on the `purge_revisions` management command in https://github.com/wagtail/wagtail/pull/10619#pullrequestreview-1528373346. We considered it out of scope to fix the issue in that PR since Wagtail has always worked like this. Hence I'm raising it again separately in this issue.

### Is your proposal related to a problem?

Wagtail provides a `purge_revisions` management command to delete stale `Revision` objects from the database. This command is aware of revisions that are still in use by Wagtail's workflow feature and avoids deleting those revisions ([link to source](https://github.com/wagtail/wagtail/blob/3d0282573d0d23f9eabe2fb6427e58b01f79ffe0/wagtail/management/commands/purge_revisions.py#L61-L66)).

However, it is somewhat common for third-party packages to rely on a specific revision. [wagtail-ab-testing](https://github.com/wagtail-nest/wagtail-ab-testing/blob/d04bd0032c4335a839c781851b12fe16037950de/wagtail_ab_testing/models.py#L76) is a real world example of this. It maintains a ForeignKey to a specific revision to track an AB experiment. If `purge_revisions` were to remove a revision in use by wagtail-ab-testing, experiment data loss would occur.

Data loss can be prevented by specifying the ForeignKey with `on_delete=models.PROTECT` but `purge_revisions` wouldn't be able to handle this and crash with a `ProtectedError`. This means third-party packages have no recourse: either they allow the revisions they depend on to be deleted or they mark their relations as protected and break Wagtails `purge_revisions` command.

### Describe the solution you'd like

I propose `purge_revisions` gracefully recovers from `ProtectedError` when trying to delete a revision. I also propose we document that relations to a `Revision` should be marked with `on_delete=models.PROTECT` if data loss is not desirable.

I also feel like the `purge_revisions` command should report the number of revisions that it couldn't delete because they were protected.

### Describe alternatives you've considered

An alternative mechanism I considered would be an API that allows packages to inform Wagtail which revisions they still use. The `purge_revisions` command can take this information into account and avoid revisions still in use.

This is a more complex solution with no apparent benefit over the solution described above.

### PR DESCRIPTION
**Pull request by NXPY123**
Fixes #10678
Rework of previous PR #10759

Added a test to ensure Revision instances having a ForeignKey with `on_delete=models.PROTECT` pointing to it are not purged while ensuring the other Revisions are purged without throwing an error.

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [x] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

File: wagtail/tests/test_management_commands.py
Comment (laymonage) [inline wagtail/tests/test_management_commands.py:726]: We need to call `save_revision()` twice and check that both revisions (especially the first revision) are not purged, because otherwise the revision will not be removed not because it's protected, but rather because it's the latest revision.

File: wagtail/tests/test_management_commands.py
Comment (laymonage) [inline wagtail/tests/test_management_commands.py:736]: There's no need to attach the latest revision to `PurgeRevisionsProtectedTestModel` as otherwise it's ambiguous whether the test ensures it's not removed due to it being the latest revision or due to `PROTECT`.

```suggestion
```

File: wagtail/tests/test_management_commands.py
Comment (laymonage) [inline wagtail/tests/test_management_commands.py:741]:
```suggestion
        # Latest revision is not deleted
        self.assertRevisionExists(revision_latest)
```

File: wagtail/tests/test_management_commands.py
Comment (laymonage) [inline wagtail/tests/test_management_commands.py:742]:
```suggestion
        # Any other revisions are deleted
        self.assertRevisionNotExists(revision_purged)
```

File: docs/reference/pages/model_reference.md
Comment (laymonage) [inline docs/reference/pages/model_reference.md:765]: There's a lot of repetitions in these paragraphs and they read a bit AI-generated to me. Also, instead of referring to "Wagtail implementors" or "administrators", just use "you" because this feature is intended for developers and we're writing developer docs.

That said, I don't think this explanation should be in the `Revision` model reference part of the documentation. I'd suggest moving this to the documentation for the `purge_revisions` command: https://github.com/wagtail/wagtail/blob/a2c9e9ab2cf3e4f0808c340f8df94882de82b5f3/docs/reference/management_commands.md?plain=1#L45

And add a new paragraph/sentence after the bullet points in the `Revision` brief to link to the `purge_revisions` docs:

https://github.com/wagtail/wagtail/blob/a2c9e9ab2cf3e4f0808c340f8df94882de82b5f3/docs/reference/pages/model_reference.md?plain=1#L676-L682

Comment (NXPY123) [reply docs/reference/pages/model_reference.md:765]: Should I remove the entire thing from the pages reference or only parts of it? It did seem kinda out of place there.

Comment (laymonage) [reply docs/reference/pages/model_reference.md:765]: I'd suggest removing it entirely and write a sentence e.g.

```markdown
You can use the [`purge_revisions`](purge_revisions) command to delete old revisions that are no longer in use.
```

to the `Revision` brief I mentioned above.

Comment (NXPY123) [reply docs/reference/pages/model_reference.md]: I've made changes accordingly in [f237d2e](https://github.com/wagtail/wagtail/pull/10961/commits/f237d2e2944784feea8d6a8576e8ad4c17c38e16). Can you please verify them?

Comment (NXPY123) [reply docs/reference/pages/model_reference.md]: I'm not sure if this:
```
[`purge_revisions`](https://github.com/wagtail/wagtail/blob/a2c9e9ab2cf3e4f0808c340f8df94882de82b5f3/docs/reference/management_commands.md?plain=1#L45)
```
is the correct way to link to the purge_revisions subheading in the management_commands.md

Comment (laymonage) [reply docs/reference/pages/model_reference.md]: You can copy-paste the block above as-is and the link should work because we've added an explicit target for it:
https://github.com/wagtail/wagtail/blob/953c980976231d6d2e1d26a58595710dcf10cc71/docs/reference/management_commands.md?plain=1#L43-L45

https://myst-parser.readthedocs.io/en/latest/syntax/cross-referencing.html

Run `make livehtml` inside the `docs` directory to build the docs locally.

Comment (NXPY123) [reply docs/reference/pages/model_reference.md]: I was wondering what that was for. That's a pretty cool functionality!

Comment (NXPY123) [reply docs/reference/pages/model_reference.md]: Can you please verify the changes in [1a173af](https://github.com/wagtail/wagtail/pull/10961/commits/1a173af7277676f8bed4548f9515c99b0cfbe808)?

File: docs/reference/management_commands.md
Comment (laymonage) [inline docs/reference/management_commands.md:79]: This documentation is written not within an `eval-rst` block, so using the RST syntax is not going to work (it has to be Markdown+MyST). You can see that this new section breaks the rest of the documentation for the management commands: https://wagtail--10961.org.readthedocs.build/en/10961/reference/management_commands.html#purge-revisions

Please build the docs locally by running `make html` or `make livehtml` in the `docs` directory before committing your changes. Make sure that the changes are rendered correctly.

For the content itself, we don't need to explain what the command does in here, as it has already been explained in the paragraphs above. Also, I think `ForeignKey` and `on_delete` are pretty standard in Django, so we don't really have to add an example (or at least not with a very generic `MyModel` like this). Let's opt for a simpler paragraph after the first paragraph.

```suggestion
```

File: docs/reference/management_commands.md
Comment (laymonage) [inline docs/reference/management_commands.md:56]: It seems to read better if we add this before the `pages` and `non-pages` option, e.g.

````suggestion
To prevent deleting important revisions when they become stale, you can refer to such revisions in a model using a `ForeignKey` with {attr}`on_delete=models.PROTECT <django.db.models.PROTECT>`.

```{versionadded} 5.2
Support for respecting `on_delete=models.PROTECT` is added.
```

If the `pages` argument is supplied, only revisions of page models will be deleted. If the `non-pages` argument is supplied, only revisions of non-page models will be deleted. If both or neither arguments are supplied, revisions of all models will be deleted.
````

Comment (Stormheg) [reply docs/reference/management_commands.md:56]: I'm 💯 behind this suggestion!

File: wagtail/management/commands/purge_revisions.py
Comment (Stormheg) [inline wagtail/management/commands/purge_revisions.py:94]: **Suggestion:** the code writing to `stdout` should probably be moved up and replace the existing code that is writing to `stdout` here: https://github.com/wagtail/wagtail/blob/1a173af7277676f8bed4548f9515c99b0cfbe808/wagtail/management/commands/purge_revisions.py#L39-L46

As it stands now, this would output duplicate messages which I don't think is desirable.

**Suggestion:** this should use Django's `self.stdout` proxy, not write to `stdout` directly. See the note about this in the Django docmentation: https://docs.djangoproject.com/en/4.2/howto/custom-management-commands/

**Suggestion:** in addition to the above, I believe the current code will write to stdout without a newline (example: `Deleted 10 revisions. Ignored 2 revisions due to ProtectedError`). I feel a newline would make sense, so let's call `write` twice and have it handle the newline for us.

**Suggestion:** finally, may I suggest we tweak the wording a bit? The intent of this suggestion is to make the message more understandable if the person seeing this message is not familiar with ProtectedError and what it signifies.

```suggestion
    self.stdout.write(f"Deleted {deleted_revisions_count} revisions. ")
    self.stdout.write(f"Ignored {protected_error_count} revisions because one or more protected relations exist that prevent deletion.")
```

**Review by Stormheg — CHANGES_REQUESTED** [review]
Hey @NXPY123, thanks for working on this! Sorry it took me so long to review, glad Sage was able to provide you with feedback in the meantime 😅

I hope you don't experience our suggestions as harsh critique, that is not my or Sage's intent. Our goal is to help make your contribution real high quality.

Do you think you are able to incorporate our feedback? Let us know, feel free to reply on a suggestion or ask in a comment if something is unclear.

Again, thank you for your time contributing 🌈

File: docs/reference/management_commands.md
Comment (NXPY123) [reply docs/reference/management_commands.md:79]: Okay, I'll build the changes and check. I'll figure out how to do it and make the fix.

**Comment by NXPY123** [comment]
Thanks for looking into the PR Stormheg! I don't think the reviews are harsh in any way and I have to thank you and Sage for taking the time to periodically review them 😅. The feedback has helped me learn more and understand some things better. I'll look into the new feedback and make changes accordingly 🙌

File: wagtail/management/commands/purge_revisions.py
Comment (NXPY123) [reply wagtail/management/commands/purge_revisions.py]: I was confused there because the PR that worked on it before added sys.stdout and the code seemed to return the deleted_revisions_count. I didn't realise the handle function was just above it 😅

Comment (NXPY123) [reply wagtail/management/commands/purge_revisions.py]: I've made changes in [ee006fc](https://github.com/wagtail/wagtail/pull/10961/commits/ee006fc18f60a32339af2b5ac10c635e59589045) and [9649d7f](https://github.com/wagtail/wagtail/pull/10961/commits/9649d7f79f63ae1747e017194b163a2a4bd12550). Can you please verify them?

Comment (NXPY123) [reply wagtail/management/commands/purge_revisions.py]: Also so far I've been taking suggestions you guys gave and making corresponding changes in my local branch and pushing them to the origin. Should I just do `commit suggestion` instead? Does that have any additional benefits compared to the way I've been doing it?

Comment (lb-) [reply wagtail/management/commands/purge_revisions.py]: Commit suggestion is fine - just remember to pull the remote to your local branch before making other changes locally. 

Normally, when merging, the core team will add a change log entry / release notes and then squash/rebase anyway. So having a few extra commits in the branch is fine.

Comment (lb-) [reply wagtail/management/commands/purge_revisions.py]: Do be sure to check the suggestions work of course though.

Comment (NXPY123) [reply wagtail/management/commands/purge_revisions.py]: Sure I'll do that 🙌

**Comment by lb-** [comment]
@Stormheg - it looks like changes have been made, would you be OK to do another review?

File: wagtail/management/commands/purge_revisions.py
Comment (Stormheg) [inline wagtail/management/commands/purge_revisions.py:50]: Perfect! ⭐️

**Review by Stormheg — APPROVED** [review]
Thanks @NXPY123!

This looks great to me. Let's merge 🚀

**Comment by Stormheg** [comment]
@NXPY123 I've credited you as 'Neeraj P Yetheendran' (from your GitHub profile).

I see you are already listed in [contributors.md](https://github.com/wagtail/wagtail/blob/6aa5b0aae038a730e1c346af236b8d26dc8e5b87/CONTRIBUTORS.md?plain=1#L735) as 'Neeraj Yetheendran'. Hope that small difference is okay!

**Pull request merged**
